### ISSUE TITLE
Weird artifacts from instance segmentation prediction from a panoptic model on a Raspberry Pi 4

### ISSUE THREAD

**Issue by juan57** [facebookresearch/detectron2#2952]
I ran a [R50-FPN](https://github.com/facebookresearch/detectron2/blob/master/MODEL_ZOO.md#coco-panoptic-segmentation-baselines-with-panoptic-fpn) on a image from the coco dataset using [demo.py](https://github.com/facebookresearch/detectron2/blob/master/demo/demo.py) on the **Raspberry Pi 4 (cpu)** and got some weird visual artifacts.

![prediction](https://user-images.githubusercontent.com/31523812/115603796-96074480-a2ae-11eb-8dc8-60976397ec20.png)

Here is the original image, predictions on the Pi vs a PC for reference.
| Original Image        | Model Prediction on Raspberry Pi 4 (cpu)          | Model Prediction on a PC (cpu) |
| ------------- |-------------|-------------|
| ![input1](https://user-images.githubusercontent.com/31523812/115605116-2003dd00-a2b0-11eb-8365-de375e8bce82.jpg)      | ![prediction](https://user-images.githubusercontent.com/31523812/115603796-96074480-a2ae-11eb-8dc8-60976397ec20.png) |![prediction](https://user-images.githubusercontent.com/31523812/115605296-593c4d00-a2b0-11eb-9695-42454117ff03.png) |

While looking into the Raspberry Pi 4 [model outputs](https://detectron2.readthedocs.io/en/latest/tutorials/models.html#model-output-format), I found that the problem is in the instance mask predictions portion of the panoptic model (for example, here is the binary mask model prediction for the horse instance object).
```
predictions = self.predictor(image)
bin_mask = predictions["instances"][0].to("cpu").pred_masks.numpy()[0,:,:] # Plotted this using plt.imsave
```
| Model Prediction on Raspberry Pi 4 (cpu)          | Model Prediction on a PC (cpu) |
|-------------|-------------|
| ![test_correct](https://user-images.githubusercontent.com/31523812/115604034-e4b4de80-a2ae-11eb-9810-d169cd147d53.png) | ![test_correct_2](https://user-images.githubusercontent.com/31523812/115606692-06639500-a2b2-11eb-8cbe-6c02c06d4341.png)  |

The code runs normally on a Ubuntu 20.04 PC workstation, but not when I run it on the [Raspberry Pi 4](https://www.raspberrypi.org/products/raspberry-pi-4-model-b/specifications/). Could this be due to a difference in architecture? (e.g. ARM vs x86) If so, is there any work arounds for this issue?

**Edit:** I have tried running the same code on the [Jetson Nano](https://developer.nvidia.com/embedded/jetson-nano) and it works fine, so I'm not sure what's going on.

Any help would be greatly appreciated!

## Instructions To Reproduce the Issue:
**On a raspberry pi 4**
1. Install raspberry pi 4 image with torch 1.8 preinstalled (https://github.com/Qengineering/RPi-image)
2. Install Detectron2 using command below according to [installation instructions](https://detectron2.readthedocs.io/en/latest/tutorials/install.html)
```
python -m pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.8/index.html
```
3. Download detectron2 repo
```git clone https://github.com/facebookresearch/detectron2.git```
4. cd to detectron2/demo
5. (download image)[https://user-images.githubusercontent.com/31523812/115605116-2003dd00-a2b0-11eb-8365-de375e8bce82.jpg] and rename as input1.jpg
6. Run the command:
```
python demo.py --config-file ../configs/COCO-PanopticSegmentation/panoptic_fpn_R_50_1x.yaml \
  --input input1.jpg --opts MODEL.WEIGHTS detectron2://COCO-PanopticSegmentation/panoptic_fpn_R_50_1x/139514544/model_final_dbfeb4.pkl
```

## Expected behavior:
| Expected Behaviour | Model Prediction on Raspberry Pi 4 (what I get)          |
|-------------|-------------|
| ![prediction](https://user-images.githubusercontent.com/31523812/115605296-593c4d00-a2b0-11eb-9695-42454117ff03.png) | ![prediction](https://user-images.githubusercontent.com/31523812/115603796-96074480-a2ae-11eb-8dc8-60976397ec20.png) |

## Environment:

- Raspberry Pi 4
- Image: https://github.com/Qengineering/RPi-image
- Installed Detectron2 via ```python -m pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.8/index.html```

**Comment by ppwwyyxx** [comment]
That's interesting to know. However we're unable to investigate further since it only reproduces on certain device. My best guess is that some ops in pytorch, or in detectron2 does not compute correct outputs on RPi. I would recommend comparing the intermediate outputs to find out why this happens. 

Since the box prediction is still correct, I would bet on the `paste_masks_in_image` function being the culprit. If this is the case, it should be simplified further and reported as a pytorch bug

**Comment by juan57** [comment]
Thanks for the info! Since it worked on a jetson nano, I decided to use that instead of the RPi. Hopefully someone else can implement a fix

**Comment by amrahsmaytas** [comment]
> I ran a [R50-FPN](https://github.com/facebookresearch/detectron2/blob/master/MODEL_ZOO.md#coco-panoptic-segmentation-baselines-with-panoptic-fpn) on a image from the coco dataset using [demo.py](https://github.com/facebookresearch/detectron2/blob/master/demo/demo.py) on the **Raspberry Pi 4 (cpu)** and got some weird visual artifacts.
> 
> ![prediction](https://user-images.githubusercontent.com/31523812/115603796-96074480-a2ae-11eb-8dc8-60976397ec20.png)
> 
> Here is the original image, predictions on the Pi vs a PC for reference.
> 
> Original Image	Model Prediction on Raspberry Pi 4 (cpu)	Model Prediction on a PC (cpu)
> ![input1](https://user-images.githubusercontent.com/31523812/115605116-2003dd00-a2b0-11eb-8365-de375e8bce82.jpg)	![prediction](https://user-images.githubusercontent.com/31523812/115603796-96074480-a2ae-11eb-8dc8-60976397ec20.png)	![prediction](https://user-images.githubusercontent.com/31523812/115605296-593c4d00-a2b0-11eb-9695-42454117ff03.png)
> While looking into the Raspberry Pi 4 [model outputs](https://detectron2.readthedocs.io/en/latest/tutorials/models.html#model-output-format), I found that the problem is in the instance mask predictions portion of the panoptic model (for example, here is the binary mask model prediction for the horse instance object).
> 
> ```
> predictions = self.predictor(image)
> bin_mask = predictions["instances"][0].to("cpu").pred_masks.numpy()[0,:,:] # Plotted this using plt.imsave
> ```
> 
> Model Prediction on Raspberry Pi 4 (cpu)	Model Prediction on a PC (cpu)
> ![test_correct](https://user-images.githubusercontent.com/31523812/115604034-e4b4de80-a2ae-11eb-9810-d169cd147d53.png)	![test_correct_2](https://user-images.githubusercontent.com/31523812/115606692-06639500-a2b2-11eb-8cbe-6c02c06d4341.png)
> The code runs normally on a Ubuntu 20.04 PC workstation, but not when I run it on the [Raspberry Pi 4](https://www.raspberrypi.org/products/raspberry-pi-4-model-b/specifications/). Could this be due to a difference in architecture? (e.g. ARM vs x86) If so, is there any work arounds for this issue?
> 
> **Edit:** I have tried running the same code on the [Jetson Nano](https://developer.nvidia.com/embedded/jetson-nano) and it works fine, so I'm not sure what's going on.
> 
> Any help would be greatly appreciated!
> 
> ## Instructions To Reproduce the Issue:
> **On a raspberry pi 4**
> 
> 1. Install raspberry pi 4 image with torch 1.8 preinstalled (https://github.com/Qengineering/RPi-image)
> 2. Install Detectron2 using command below according to [installation instructions](https://detectron2.readthedocs.io/en/latest/tutorials/install.html)
> 
> ```
> python -m pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.8/index.html
> ```
> 
> 1. Download detectron2 repo
>    `git clone https://github.com/facebookresearch/detectron2.git`
> 2. cd to detectron2/demo
> 3. (download image)[https://user-images.githubusercontent.com/31523812/115605116-2003dd00-a2b0-11eb-8365-de375e8bce82.jpg] and rename as input1.jpg
> 4. Run the command:
> 
> ```
> python demo.py --config-file ../configs/COCO-PanopticSegmentation/panoptic_fpn_R_50_1x.yaml \
>   --input input1.jpg --opts MODEL.WEIGHTS detectron2://COCO-PanopticSegmentation/panoptic_fpn_R_50_1x/139514544/model_final_dbfeb4.pkl
> ```
> 
> ## Expected behavior:
> Expected Behaviour	Model Prediction on Raspberry Pi 4 (what I get)
> ![prediction](https://user-images.githubusercontent.com/31523812/115605296-593c4d00-a2b0-11eb-9695-42454117ff03.png)	![prediction](https://user-images.githubusercontent.com/31523812/115603796-96074480-a2ae-11eb-8dc8-60976397ec20.png)
> ## Environment:
> * Raspberry Pi 4
> * Image: https://github.com/Qengineering/RPi-image
> * Installed Detectron2 via `python -m pip install detectron2 -f https://dl.fbaipublicfiles.com/detectron2/wheels/cpu/torch1.8/index.html`

Hi @juan57 
I am trying to implement d2go on rpi, but i am not sure about the inference accuracy and speed on real time footage's of 5 custom classes.

can you help with the numbers, please?

I want to know whether the accuracy and speed differ in rpi from speed: 0.05 sec & given 39.25 for instance segmentation?
if yes, how much?

Thanks and warm regards,
satyam.

**Comment by juan57** [comment]
Not sure what you mean by “accuracy and speed differ by speed” - but for us, it took somewhere around 30 sec - 1 minute to run inference on a pretrained R50-FPN for a single image, and accuracy for your case I can’t really comment on without more detail.

**Comment by fapbatista** [comment]
@juan57 I got a similar problem on a custom model trained on a custom dataset. What we have in Common is three things: RPi 4 + PyTorch + the QEngineering Guide to install PyTorch and Torchvision (I had a clean Ubuntu Desktop install and installed in miniconda using their wheels)

**Comment by fapbatista** [comment]
@juan57 The issue seems to be the QEngineering wheels of Pytorch. Compiling PyTorch and Torch Vision from scratch in my RPi 4B did the trick. I get precise masks now. Hope it helps.

**Comment by juan57** [comment]
@fapbatista  That's great to know! Thanks!

**Comment by XiaoBaoBao719** [comment]
Hey @juan57 I'm also working on getting detectron2 on my Raspberry Pi 4. I was curious how much RAM your model has? I'm using the 2 GB version and unfortunately my Pi is freezing while unpacking detectron2. I've been closely following the steps that you provided as well as on the detectron2 documentation page.

**Comment by juan57** [comment]
It's been a few months since I worked on the project, but I believe we used a Raspberry Pi 4 with 8 GB ram. Not sure, but 2 GB seems a bit small.

**Comment by sourabmaity** [comment]
@juan57 are you running your model in Gpu or in Cpu<br>
I am using detectron2 but in CPU so its predicting so slow (like 15 s per prediction) so how can I use GPU in raspberry Pi 4
